query-planner-js/src/buildPlan.ts

```diff
@@ -1522,6 +1539,34 @@ class FetchDependencyGraph {
     }
   }
 
+  private removeUselessGroups(group: FetchGroup) {
+    // Recursing first, this makes it a bit easier to reason about.
+    for (const child of group.children()) {
+      this.removeUselessGroups(child);
+    }
+
+    // If a group is such that everything is fetches is already included in the inputs, then
+    // this group does useless fetches and can be removed.
+    if (group.inputs && group.inputs.contains(group.selection)) {
+      // In general, removing a group is a bit tricky because we need to deal with the fact
+      // that the group can have multiple parents and children and no break the "path in parent"
+      // in all those cases. To keep thing relatively easily, we only handle the following
+      // cases (other cases will remain non-optimal, but hopefully this handle all the cases
+      // we care about in practice):
+      //   1. if the group has no children. In which case we can just remove it with no ceremony.
+      //   2. if the group has only a single parent and we have a path to that parent.
+      if (group.children().length === 0) {
+        this.remove(group);
+      } else {
+        const parents = group.parents();
+        const parent = parents[0];
+        if (parents.length === 1 && parent.path) {
```

Comment:
Just for my understanding, what are the conditions under which a parent wouldn't have a path?

Reply:
There is some cases where the `path` doesn't really make sense at I explained in [this comment](https://github.com/apollographql/federation/blob/main/query-planner-js/src/buildPlan.ts#L603-L607). But more concretely, it show in code paths at least [here](https://github.com/apollographql/federation/blob/main/query-planner-js/src/buildPlan.ts#L3274) and [here](https://github.com/apollographql/federation/blob/main/query-planner-js/src/buildPlan.ts#L3319).